### PR TITLE
Add Android SDK alpha version

### PR DESCRIPTION
**Pull request by danielnugraha**
### Description

Existing android examples have similarities that can be extracted into its own client SDK, and also it lacks the new gRPC-rere implementation.

### Explanation

1. Implemented the initial alpha version of Android SDK, which includes Typing, Serde, TaskHandler, MessageHandler, Client, gRPC (normal and rere).
2. Protobuf generation is based on `src/proto`.
3. Package the SDK into an Android library.

### Checklist

- [ ] Implement proposed change
- [ ] Write tests
- [ ] Update [documentation](https://flower.dev/docs/writing-documentation.html)
- [ ] Update [changelog](https://github.com/adap/flower/blob/main/doc/source/changelog.rst)
- [ ] Make CI checks pass
- [ ] Ping maintainers on [Slack](https://flower.dev/join-slack/) (channel `#contributions`)

### Any other comments?

gRPC-ReRe implementation is using producer and consumer principle for managing taskIns request and response.
   - Initial PullTaskInsRequest message will be made via calling function request()
   - Receive function is called, it acts as a consumer that monitors the requestChannel, if there is a request in the channel, it immediately sends the request to the server. If it receives empty response, it will wait for 3 seconds and call request() to put PullTaskInsRequest message in the channel, else it emits the response as a flow.
   - In startGRPCReRe function, the emitted response will be handled by handle() function and its corresponding taskRes will be sent to the server. The function request() will be called once again if keepGoing is true, else deleteNode() is called and the channel will be closed.

### CODE REVIEW

File: .github/workflows/android-release.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+        # Runs upload, and then closes & releases the repository
+      - name: Publish to MavenCentral
+        run: ./gradlew publishAllPublicationsToMavenCentral --no-configuration-cache
+        env:
+          ORG_GRADLE_PROJECT_mavenCentralUsername: ${{ secrets.OSSRH_USERNAME }}
+          ORG_GRADLE_PROJECT_mavenCentralPassword: ${{ secrets.OSSRH_PASSWORD }}
+          ORG_GRADLE_PROJECT_signingInMemoryKeyId: ${{ secrets.SIGNING_KEY_ID }}
+          ORG_GRADLE_PROJECT_signingInMemoryKeyPassword: ${{ secrets.SIGNING_PASSWORD }}
+          ORG_GRADLE_PROJECT_signingInMemoryKey: ${{ secrets.SIGNING_KEY }}
```
Comment (tanertopal) [inline .github/workflows/android-release.yml:32]:
```suggestion
          ORG_GRADLE_PROJECT_signingInMemoryKey: ${{ secrets.SIGNING_KEY }}

```

File: src/kotlin/.gitignore [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+/.idea/modules.xml
+/.idea/workspace.xml
+/.idea/navEditor.xml
+/.idea/assetWizardSettings.xml
+.DS_Store
+/build
+/captures
+.externalNativeBuild
+.cxx
+local.properties
```
Comment (tanertopal) [inline src/kotlin/.gitignore:15]: I am wondering if we should move this to the top-level .gitignore file. Wdyt?

File: src/kotlin/flwr/.gitignore [hunk 3]
```diff
@@ -0,0 +1 @@
+/build
```
Comment (tanertopal) [inline src/kotlin/flwr/.gitignore:1]: Same as the previous comment on `src/kotlin/.gitignore`

File: src/kotlin/.gitignore (hunk 2 above)
Comment (danielnugraha) [reply src/kotlin/.gitignore:15]: It's good idea, I'll move it to top level

File: src/kotlin/flwr/src/main/java/dev/flower/android/gRPC.kt [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+import flwr.proto.NodeOuterClass.Node
+import flwr.proto.TaskOuterClass.TaskIns
+import flwr.proto.TaskOuterClass.TaskRes
+import flwr.proto.Transport.ServerMessage
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.flow
+
+internal class FlowerGRPC
```
Comment (danieljanes) [inline src/kotlin/flwr/src/main/java/dev/flower/android/gRPC.kt:27]: We usually `CamelCase` abbreviations like gRPC or HTTP (i.e., `Grpc` / `Http`):
```suggestion
internal class FlowerGrpc
```

File: src/kotlin/flwr/src/main/java/dev/flower/android/gRPC.kt [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+ * Start a Flower client node which connects to a Flower server.
+ *
+ * @param serverAddress The IPv4 or IPv6 address of the server. If the Flower server runs on the
+ * same machine on port 8080, then server_address would be “[::]:8080”.
+ * @param useTLS Whether to use TLS to connect to the Flower server.
+ * @param client The Flower client implementation.
+ */
+suspend fun startClient(
+    serverAddress: String,
+    useTLS: Boolean,
```
Comment (danieljanes) [inline src/kotlin/flwr/src/main/java/dev/flower/android/gRPC.kt:71]:
```suggestion
    useTls: Boolean,
```

File: src/kotlin/flwr/src/main/java/dev/flower/android/gRPC.kt [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+                    delay(3000)
+                    request(requestChannel, node)
+                } else {
+                    emit(response)
+                }
+            }
+        }
+    }
+
+    suspend fun startGRPCReRe() {
```
Comment (danieljanes) [inline src/kotlin/flwr/src/main/java/dev/flower/android/gRPC.kt:183]:
```suggestion
    suspend fun startGrpcRere() {
```

**Review by danieljanes — CHANGES_REQUESTED** [review]

File: src/kotlin/gradle.properties [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+SONATYPE_HOST=S01
+RELEASE_SIGNING_ENABLED=true
+
+GROUP=dev.flower
+POM_ARTIFACT_ID=flwr-android
+VERSION_NAME=0.0.2
+
+POM_NAME=Flwr Android
+POM_DESCRIPTION=Friendly Federated Learning Framework
```
Comment (danieljanes) [inline src/kotlin/gradle.properties:33]:
```suggestion
POM_NAME=Flower Android
POM_DESCRIPTION=A Friendly Federated Learning Framework
```

File: src/kotlin/gradle.properties [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+# Enables namespacing of each library's R class so that its R class includes only the
+# resources declared in the library itself and none from the library's dependencies,
+# thereby reducing the size of the R class for that library
+android.nonTransitiveRClass=true
+
+SONATYPE_HOST=S01
+RELEASE_SIGNING_ENABLED=true
+
+GROUP=dev.flower
+POM_ARTIFACT_ID=flwr-android
```
Comment (danieljanes) [inline src/kotlin/gradle.properties:29]: Can this be just `flwr`, or `flower`?

File: src/kotlin/flwr/src/main/java/dev/flower/android/gRPC.kt [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+                    val request = PushTaskResRequest.newBuilder().addTaskResList(taskRes).build()
+                    asyncStub.pushTaskRes(request, null)
+                }
+                state[KEYTASKINS] = null
+            }
+        }
+    }
+}
+
+suspend fun createFlowerReRe(
```
Comment (danieljanes) [inline src/kotlin/flwr/src/main/java/dev/flower/android/gRPC.kt:230]: Let's change `ReRe` everywhere to `Rere`:
```suggestion
suspend fun createFlowerRere(
```

File: src/kotlin/flwr/src/main/java/dev/flower/android/gRPC.kt [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+        channelBuilder.usePlaintext()
+    }
+    return withContext(Dispatchers.IO) {
+        channelBuilder.build()
+    }
+}
+
+const val HUNDRED_MEBIBYTE = 100 * 1024 * 1024
+
+internal class FlwrReRe
```
Comment (danieljanes) [inline src/kotlin/flwr/src/main/java/dev/flower/android/gRPC.kt:90]: Same as below

File: src/kotlin/flwr/src/main/java/dev/flower/android/gRPC.kt [hunk 11]
```diff
@@ -0,0 +1,236 @@
+package dev.flower.android
```
Comment (danieljanes) [inline src/kotlin/flwr/src/main/java/dev/flower/android/gRPC.kt:1]: Should we name the entire file `Grpc.kt` for consistency?

**Review by danieljanes — CHANGES_REQUESTED** [review]

File: src/kotlin/gradle.properties (hunk 8 above)
Comment (danielnugraha) [reply src/kotlin/gradle.properties:29]: Group should be equal to our website flower.dev reversed

Comment (danielnugraha) [reply src/kotlin/gradle.properties:29]: Fixed

**Review by danieljanes — APPROVED** [review]

**Pull request merged**